language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -55,14 +50,46 @@ optionsParserInfo = info (optionsParser <**> helper)
 main :: IO ()
 main = do
     opts@Options{..} <- execParser optionsParserInfo
-    dar <- B.readFile optInputDar
-    let archive = Zip.toArchive $ BSL.fromStrict dar
-    dalfs <- either fail pure $ DAR.readDalfs archive
-    DAR.DalfManifest{packageName, ..} <- either fail pure $ DAR.readDalfManifest archive
-    packageName <- pure $ optMainPackageName <|> packageName
-    forM_ ((DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)) $ \(dalf, mbPkgName) -> do
-        (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
-        daml2ts opts pkgId pkg mbPkgName
+    foldM_ (processDar opts) Map.empty optInputDars
```

Comment:
I always get confused by nested folds and I think this is also sensitive to the order which is probably not intended: If I first see `a.dar` which has a dependency on `$pkgid1` and afterwards `b.dar` where the main dalf has `$pkgid1` I’m not going to pick up the name for `$pkgid1`. If I put `b.dar` first I will pick up the name. 

I would suggest to do this in two steps instead:

1. Take a list of DARs and build up a list of `Map.Map PackageId ([String], Archive)`. This basically consist of a function `FilePath -> Map.Map PackageId ([String], Archive)` that you apply to each DAR and then throw something like `Map.unionsWith (\(names1, archive), (names2, _) -> (names1 <> names2, archive)` at it to merge them (a bit more complex if you want to detect duplicate package names).
2. Check for errors, if there are none iterate over the map and generate typescript.

Reply:
> I always get confused by nested folds and I think this is also sensitive to the order which is probably not intended: If I first see a.dar which has a dependency on $pkgid1 and afterwards b.dar where the main dalf has $pkgid1 I’m not going to pick up the name for $pkgid1. If I put b.dar first I will pick up the name.

I'm not sure about that.

If you see `a.dar` and it depends on `$pkgid1`, then ts will be generated for `$pkgid1` (against its hash). If later you see `b.dar` where the main dalf has `$pkgid1`, it will generate again this time against its package name (unless the package name is missing in which case it won't regenerate).

Reply:
Oh I see, you will actually call `daml2ts` twice. That’s not what I would have expected. I would have expected that we always generate once per package id. If we can somehow infer a name for that package id we use it (assuming it’s unique, otherwise we error out). If there is no name we generate it without a name.

Reply:
Cool. I hear you. This might not be the end game but yes, the scheme implemented here in some cases will result in some duplicated packages (1.a  of https://github.com/digital-asset/daml/issues/4585). Thanks for looking at that somewhat tricky logic Moritz!

Reply:
I could live with the duplicates for now but ultimately they'll have to go. What @cocreature describes would solve that problem _and_ most likely be easier to understand since it separates the conflict detection from the code generation.

I would suggest to build an intermediate structure of type `Map PackageId (Archive, Maybe String)`, where the `Maybe String` part is an optional name for a package. If we see a second name, we can fail immediately. We can build this intermediate structure in one big `foldl'` over _all_ DALFs and their potential names.